Update v2 > v3 migration guide (#10728)
traefik/traefik: docs/content/migration/v2-to-v3.md

How to Migrate from Traefik v2 to Traefik v3.
{: .subtitle }

> This guide shares 4 steps to progressively migrate from Traefik v2 to v3:

Reviewer: ```suggestion
With Traefik v3, we are introducing a streamlined transition process from v2. Minimal breaking changes have been made on specific options on the static configuration, and we are ensuring backward compatibility with v2 syntax on the dynamic configuration. This will offer a gradual path for adopting the v3 syntax, allowing users to progressively migrate their Kubernetes ingress resources, Docker labels, etc. to the new format.

Here are the steps to progressively migrate from Traefik v2 to v3:
```

How to Migrate from Traefik v2 to Traefik v3.
{: .subtitle }

> This guide shows you how to progressively migrate from Traefik v2 to v3.

Reviewer: In the intro, we should list the steps in which we deep dive later.

Here are the steps to progressively migrate from Traefik v2 to v3:

1. [Prepare configurations and test v3](#step-1-prepare-configurations-and-test-v3)
1. [Migrate production instances to Traefik v3](#step-2-migrate-production-instances-to-traefik-v3)
1. [Progressively migrate dynamic configuration](#step-3-progressively-migrate-dynamic-configuration)

## Step 1: Prepare Configurations and Test v3

Check the changes in [static configurations](./v2-to-v3-details.md#static-configuration-changes "Link to static configuration changes") and [operations](./v2-to-v3-details.md#operations-changes "Link to operations changes") brought by Traefik v3.
Modify your configurations accordingly.

Then, add the following snippet to the static configuration:

```yaml
# static configuration
core:
  defaultRuleSyntax: v2
```

> The snippet makes static configuration use the default [v2 syntax](https://doc.traefik.io/traefik/v3.0/migration/v2-to-v3/?ref=traefik.io#configure-the-default-syntax-in-static-configuration "Link to configure default syntax in static config").

Reviewer: ```suggestion
This snippet in the static configuration makes the [v2 format](https://doc.traefik.io/traefik/v3.0/migration/v2-to-v3/?ref=traefik.io#configure-the-default-syntax-in-static-configuration "Link to configure default syntax in static config") the default rule matchers syntax.
```

### Rancher v1 Provider

> In v3, the Rancher v1 provider has been removed because Rancher v1 is [no longer actively maintained](https://rancher.com/docs/os/v1.x/en/support/). Rancher v2 is supported as a standard Kubernetes provider.

Reviewer 2: ```suggestion
In v3, the Rancher v1 provider has been removed because Rancher v1 is [no longer actively maintained](https://rancher.com/docs/os/v1.x/en/support/).
Rancher v2 is supported as a standard Kubernetes provider.
```

All Pilot related configuration should be removed from the static configuration.

---

Once you have prepared your static configuration, add the following snippet to it:

```yaml
# static configuration
core:
  defaultRuleSyntax: v2
```

The snippet makes static configuration use the default [v2 syntax](https://doc.traefik.io/traefik/v3.0/migration/v2-to-v3/?ref=traefik.io#configure-the-default-syntax-in-static-configuration "Link to configure default syntax in static config").

Start Traefik v3 with this new configuration to test it.

!!! info
    If you don’t get any error logs while testing, you are good to go!
    Otherwise, follow the remaining migration options highlighted in the logs.

Once your Traefik test instances are starting and routing to your applications, proceed to the next step.

## Step 2: Migrate Production Instances to Traefik v3

> Use progressive migration strategy  ([Kubernetes rolling update mechanism](https://kubernetes.io/docs/tutorials/kubernetes-basics/update/update-intro/ "Link to the Kubernetes rolling update documentation"), for example) to migrate your production instances to v3.

Reviewer: ```suggestion
We strongly advise you to follow a progressive migration strategy ([Kubernetes rolling update mechanism](https://kubernetes.io/docs/tutorials/kubernetes-basics/update/update-intro/ "Link to the Kubernetes rolling update documentation"), for example) to migrate your production instances to v3.
```

!!! Warning
    Ensure you have a [real-time monitoring solution](https://traefik.io/blog/capture-traefik-metrics-for-apps-on-kubernetes-with-prometheus/ "Link to the blog on capturing Traefik metrics with Prometheus") for your ingress traffic to detect issues instantly.

During the progressive migration, monitor your ingress traffic for any errors. Be prepared to rollback to a working state in case of any issues.

!!! Info
    If you encounter any issues, leverage debug and access logs provided by Traefik to understand what went wrong and how to fix it.

Once every Traefik instance is updated, you will be on Traefik v3!

> ## Step 3: Progressively Migrate Dynamic Configuration

Reviewer: This section also includes the Operations section, which is super misleading.
We should probably group _Static Configuration_ & _Operations Changes_ together in this guide. WDYT @nmengin ?

!!! info

>     You can do this later, as Traefik v3 is compatible with the v2 format for [dynamic configuration](https://doc.traefik.io/traefik/v3.0/migration/v2-to-v3/#dynamic-configuration "Link to dynamic configuration changes").

Reviewer: ```suggestion
    This step can be done later in the process, as Traefik v3 is compatible with the v2 format for [dynamic configuration](https://doc.traefik.io/traefik/v3.0/migration/v2-to-v3/#dynamic-configuration "Link to dynamic configuration changes").
```

> Check the following changes in dynamic configuration.

Reviewer: I would also add a section saying that the user should enable & follow Traefik logs during this step, as they will provide some help in case a deprecated option is in use.

### TCP LoadBalancer `terminationDelay` option

The TCP LoadBalancer `terminationDelay` option has been removed.

> This option can now be configured directly on the `TCPServersTransport` level. Please take a look at this [documentation](../routing/services/index.md#terminationdelay).

Reviewer 2: ```suggestion
This option can now be configured directly on the `TCPServersTransport` level.
Please take a look at this [documentation](../routing/services/index.md#terminationdelay).
```

By default, the `defaultRuleSyntax` static option is automatically set to `v3`, meaning that the default rule is the new one.

> ##### New V3 Syntax Notable Changes

Reviewer 3: Could you please take into account the following comment https://github.com/traefik/traefik/issues/10690#issuecomment-2145134722
